File: dashboard/android/app/src/main/java/com/attobot/dashboard/ui/screens/MemoryScreen.kt
package com.attobot.dashboard.ui.screens

import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import androidx.navigation.NavHostController
import com.attobot.dashboard.core.MemoryRow
import com.attobot.dashboard.core.timeAgo
import com.attobot.dashboard.core.truncate
import com.attobot.dashboard.state.MemoryViewModel
import com.attobot.dashboard.state.UiState
import com.attobot.dashboard.ui.components.AgentPills
import com.attobot.dashboard.ui.components.DataTable
import com.attobot.dashboard.ui.components.ErrorState
import com.attobot.dashboard.ui.components.LoadingView
import com.attobot.dashboard.ui.components.PullRefreshScreen
import com.attobot.dashboard.ui.components.TableColumn
import com.attobot.dashboard.ui.theme.accent
import com.attobot.dashboard.ui.theme.err
import com.attobot.dashboard.ui.theme.ok
import com.attobot.dashboard.ui.theme.text

@Composable
fun MemoryScreen(
    @Suppress("UNUSED_PARAMETER") navController: NavHostController,
    initialAgentId: String?,
) {
    val vm: MemoryViewModel = viewModel()
    LaunchedEffect(initialAgentId) {
        if (!initialAgentId.isNullOrEmpty()) vm.setAgentId(initialAgentId)
    }
    val state by vm.state.collectAsStateWithLifecycle()
    val agentId by vm.agentId.collectAsStateWithLifecycle()

    val s = state
    val refreshing = s is UiState.Ready && s.refreshing

    PullRefreshScreen(refreshing = refreshing, onRefresh = vm::refresh) {
        Text(
            "Memory",
            color = text,
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold,
            modifier = Modifier.padding(bottom = 12.dp),
        )
        AgentPills(agentId, vm::setAgentId)
        when (s) {
            is UiState.Loading -> LoadingView()
            is UiState.Error -> ErrorState(s.message)
            is UiState.Ready -> DataTable(rows = s.data, columns = memoryColumns())
        }
    }
}

private fun memoryColumns(): List<TableColumn<MemoryRow>> = listOf(
    TableColumn(header = "id", width = 80.dp) {
        Text("${it.id}", color = accent, fontFamily = FontFamily.Monospace, fontSize = 12.sp)
    },
    TableColumn(header = "agent", width = 70.dp) {
        Text("${it.agentId}", color = text, fontSize = 13.sp)
    },
    TableColumn(header = "content", width = 220.dp) {
        Text(
            truncate(it.content, 160),
            color = text,
            fontSize = 13.sp,
            maxLines = 3,
            overflow = TextOverflow.Ellipsis,
        )
    },
    TableColumn(header = "on", width = 40.dp) {
        val on = it.enabled == true
        Text(if (on) "●" else "○", color = if (on) ok else err, fontSize = 13.sp)
    },
    TableColumn(header = "sources", width = 80.dp) {
        Text("${it.sourceMessageIds?.size ?: 0}", color = text, fontSize = 13.sp)
    },
    TableColumn(header = "updated", width = 90.dp) {
        Text(timeAgo(it.updatedAt), color = text, fontSize = 13.sp)
    },
)

File: dashboard/ios/OpenWorldDashboard/Views/Screens/MemoryScreen.swift
import SwiftUI

/// Memory table. No poll; refreshable. AgentPills only.
struct MemoryScreen: View {
    let initialAgentId: String?
    @State private var agentId: String
    @State private var res = Resource<[MemoryRow]>()

    init(initialAgentId: String?) {
        self.initialAgentId = initialAgentId
        _agentId = State(initialValue: initialAgentId ?? "")
    }

    var body: some View {
        ScreenScroll {
            Text("Memory")
                .font(.system(size: 20, weight: .bold))
                .foregroundStyle(Theme.text)
                .padding(.bottom, 12)

            AgentPills(value: agentId, onSelect: { agentId = $0 })
                .padding(.horizontal, -16)
                .padding(.bottom, 8)

            if let err = res.error, res.value == nil {
                ErrorState(message: err)
            } else if res.isLoading && res.value == nil {
                Spinner()
            } else {
                DataTable(
                    columns: [
                        DataTableColumn<MemoryRow>(id: "id", title: "id", width: 80) { r in
                            Text("\(r.id)").font(.system(size: 12, design: .monospaced)).foregroundStyle(Theme.accent)
                        },
                        DataTableColumn<MemoryRow>(id: "agent", title: "agent", width: 70) { r in
                            Text("\(r.agentId)").font(.system(size: 13)).foregroundStyle(Theme.text)
                        },
                        DataTableColumn<MemoryRow>(id: "content", title: "content") { r in
                            Text(Format.truncate(r.content ?? "", n: 160))
                                .font(.system(size: 13))
                                .foregroundStyle(Theme.text)
                                .lineLimit(3)
                                .fixedSize(horizontal: false, vertical: true)
                        },
                        DataTableColumn<MemoryRow>(id: "enabled", title: "on", width: 40) { r in
                            Text((r.enabled ?? false) ? "●" : "○")
                                .font(.system(size: 13))
                                .foregroundStyle((r.enabled ?? false) ? Theme.ok : Theme.err)
                        },
                        DataTableColumn<MemoryRow>(id: "sources", title: "sources", width: 80) { r in
                            Text("\(r.sourceMessageIds?.count ?? 0)").font(.system(size: 13)).foregroundStyle(Theme.text)
                        },
                        DataTableColumn<MemoryRow>(id: "updated", title: "updated", width: 90) { r in
                            Text(Format.timeAgo(r.updatedAt)).font(.system(size: 13)).foregroundStyle(Theme.text)
                        },
                    ],
                    rows: res.value ?? []
                )
            }
        }
        .navigationTitle("Memory")
        .navigationBarTitleDisplayMode(.inline)
        .task(id: agentId) {
            res.start(load: {
                try await APIClient.getRows(APIClient.buildPath(
                    "/api/memory",
                    params: [("agent_id", agentId)]
                ))
            })
        }
    }
}
